### ISSUE TITLE
Suppport running Cypress in NW.js (node-webkit)

### ISSUE THREAD

**Issue by TheJaredWilcurt** [cypress-io/cypress#4662]
### Current behavior:

Shows a dropdown for Chrome/Chromium/Canary/Electron

### Desired behavior:

Show NW.js in dropdown, or have a config where you can point to the browser executable to use.

My apps runs in NW.js and rely on the globals it comes with (`window.nw`, `window.require`, etc). They cannot be ran in any other browser environment without heavy mocking.

**Comment by jennifer-shehane** [comment]
@TheJaredWilcurt There are instructions on how to launch the browser by path here: https://on.cypress.io/launching-browsers#Launching-by-a-path

And yes, if you can mock the things like `window.nw` using JavaScript, then this can be done from within Cypress as well.

**Comment by TheJaredWilcurt** [comment]
Using the provided API does not work. I get the following message:

```
**Message:** We found an invalid configuration value:

<span class="ansi-yellow-fg">Found an error while validating the `browsers` list. Expected `version` to be a non-empty string. Instead the value was: `{"name":"chrome","family":"chromium","channel":"stable","displayName":"Custom Chrome","versionRegex":{},"binary":["google-chrome","chrome","google-chrome-stable"],"info":"Loaded from node_modules/nw/nwjs/nw.exe","custom":true,"path":"node_modules/nw/nwjs/nw.exe","version":""}`</span>

**Stack trace:**
Error: We found an invalid configuration value:

[33mFound an error while validating the `browsers` list. Expected `version` to be a non-empty string. Instead the value was: `{"name":"chrome","family":"chromium","channel":"stable","displayName":"Custom Chrome","versionRegex":{},"binary":["google-chrome","chrome","google-chrome-stable"],"info":"Loaded from node_modules/nw/nwjs/nw.exe","custom":true,"path":"node_modules/nw/nwjs/nw.exe","version":""}`[39m
    at get (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\lib\errors.js:968:15)
    at Object.throwErr [as throw] (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\lib\errors.js:986:9)
    at C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\lib\config.js:303:26
    at C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\lib\config.js:96:18
    at C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\lodash\lodash.js:4925:15
    at baseForOwn (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\lodash\lodash.js:2990:24)
    at C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\lodash\lodash.js:4894:18
    at Function.forEach (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\lodash\lodash.js:9368:14)
    at validate (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\lib\config.js:86:12)
    at Object.mergeDefaults (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\lib\config.js:302:5)
    at Object.set (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\lib\config.js:238:17)
    at C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\lib\config.js:209:19
    at tryCatcher (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\util.js:16:23)
    at Promise._settlePromiseFromHandler (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:544:35)
    at Promise._settlePromise (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:604:18)
    at Promise._settlePromise0 (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:649:10)
    at Promise._settlePromises (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:729:18)
    at Promise._fulfill (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:673:18)
    at PromiseArray._resolve (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise_array.js:127:19)
    at PromiseArray._promiseFulfilled (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise_array.js:145:14)
    at PromiseArray._iterate (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise_array.js:115:31)
    at PromiseArray.init [as _init] (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise_array.js:79:10)
    at Promise._settlePromise (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:601:21)
    at Promise._settlePromise0 (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:649:10)
    at Promise._settlePromises (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:729:18)
    at Promise._fulfill (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:673:18)
    at PromiseArray._resolve (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise_array.js:127:19)
    at PromiseArray._promiseFulfilled (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise_array.js:145:14)
    at Promise._settlePromise (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:609:26)
    at Promise._settlePromise0 (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:649:10)
    at Promise._settlePromises (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:729:18)
    at Promise._fulfill (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:673:18)
    at Promise._resolveCallback (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:466:57)
    at Promise._settlePromiseFromHandler (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:559:17)
    at Promise._settlePromise (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:604:18)
    at Promise._settlePromise0 (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:649:10)
    at Promise._settlePromises (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:729:18)
    at Promise._fulfill (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:673:18)
    at Promise._settlePromise (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:617:21)
    at Promise._settlePromise0 (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:649:10)
    at Promise._settlePromises (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:729:18)
    at Promise._fulfill (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:673:18)
    at Promise._resolveCallback (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:466:57)
    at Promise._settlePromiseFromHandler (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:559:17)
    at Promise._settlePromise (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:604:18)
    at Promise._settlePromise0 (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:649:10)
    at Promise._settlePromises (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:729:18)
    at Promise._fulfill (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:673:18)
    at Promise._settlePromise (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:617:21)
    at Promise._settlePromise0 (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:649:10)
    at Promise._settlePromises (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:729:18)
    at Promise._fulfill (C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:673:18)
    at C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\nodeback.js:42:21
    at C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\jsonfile\index.js:43:5
    at C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\jsonfile\node_modules\graceful-fs\graceful-fs.js:123:16
    at C:\Users\owner\AppData\Local\Cypress\Cache\6.3.0\Cypress\resources\app\packages\server\node_modules\graceful-fs\graceful-fs.js:90:16
    at FSReqCallback.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:63:3)
```

Seems to be having trouble finding a version number I think.

**Note:** NW.js is not a normal browser, it is a modified Chromium browser with Node.js built in used for creating desktop apps.

So if anything special had to be done to make Electron compatible with Cypress, similar efforts may be required for NW.js.

Here is my minimum reproduction repo:

* https://github.com/nwutils/nw-cypress-example

**Comment by gpetrov** [comment]
it will be really awesome if we can get cypress to work with NWJS - being waiting for this for years!
Hope the cypress team pick up this issue more quickly now.

**Comment by sarendsen** [comment]
+1

**Comment by rkarroll** [comment]
+1

**Comment by gpetrov** [comment]
Maybe @jennifer-shehane can give us more insights here?

**Comment by YehielShamir** [comment]
+1

**Comment by TheJaredWilcurt** [comment]
I recently created a benchmark to compare the speed of launching/closing NW.js and Electron:

* https://github.com/TheJaredWilcurt/xpda-benchmark

On some machines the two are roughly equivalent, but on others NW.js is more than twice as fast. Having it as an option in the dropdown, could be beneficial to users (especially on Windows) who just want tests to run faster, and don't really care about which browser they're code is ran in.

This support would also essentially guarantee Cypress as the best tool for E2E within the NW.js community. I'd happily set it up in all of the boilerplates I maintain.

* https://nwutils.io

**Comment by jennifer-shehane** [comment]
@TheJaredWilcurt The first example is pointing to a file that does not exist trying to open in Cypress. `node_modules/nw/nwjs/nw.exe:chrome` Do you have an example to run with this not working?

**Comment by TheJaredWilcurt** [comment]
@jennifer-shehane `node_modules/nw/nwjs/nw.exe` does exist if you are on Windows. You will need to go into node_modules and find the binary for your own OS and change it to that. I'm not concerned with cross-platform friendliness right now. I need it to work at all first, then I'll find how to make it work on all systems.

Pointing to the file directly doesn't work because Cypress isn't coded to recognize `nw.exe` as a browser.

![cy-err](https://user-images.githubusercontent.com/4629794/125344348-2b056100-e325-11eb-987f-e60e6854e833.png)

So based on the Cypress docs from when I made that repo, I added `:chrome` to the end to instruct Cypress to treat it like a chrome browser (since NW.js is based on Chromium). Hoping that would work, it just gives a different error.

![cy-err2](https://user-images.githubusercontent.com/4629794/125345239-491f9100-e326-11eb-8b0b-beb241ceae78.png)

This is the case with it is `nw.exe:chrome` or `nw.exe:chromium`.

I've updated that repo to try a different approach based on a different section of the docs. It now loads Cypress normally, shows NW.js in the dropdown, but trying to run it crashes.
